src/client.rs

```diff
@@ -119,7 +120,57 @@ impl Client {
     where
         StreamName: AsRef<str>,
     {
-        commands::subscribe_to_stream(&self.client, stream_name, options).await
+        if !options.auto_resub {
+            return commands::subscribe_to_stream(&self.client, stream_name, options).await;
+        }
+
+        let client = self.client.clone();
+        let stream_name = stream_name.as_ref().to_string();
+        let mut options = options.clone();
+        let mut stream =
+            commands::subscribe_to_stream(&self.client, stream_name.as_str(), &options).await?;
+        let result = async_stream::stream! {
+            let mut offset = options.position;
+
+            loop {
+                match stream.try_next().await {
+                    Ok(sub_event) => {
+                        let sub_event = sub_event.expect("Event field must be defined because subs can't end");
+                        match sub_event {
+                            crate::types::SubEvent::EventAppeared(event) => {
+                                offset = crate::types::StreamPosition::Point(event.get_original_event().revision);
+                                yield Ok(crate::types::SubEvent::EventAppeared(event));
+                            }
+
+                            ignored => yield Ok(ignored),
+                        }
+                    }
+
+                    Err(e) => {
+                        error!("Regular subscription dropped: {:?}. Reconnecting...", e);
+
+                        options = options.position(offset);
+
+                        loop {
+                            match commands::subscribe_to_stream(&client, stream_name.as_str(), &options).await {
+                                Ok(new_stream) => {
+                                    stream = new_stream;
+                                    break;
+                                }
+
+                                Err(e) => {
+                                    error!("Exception when re-subscribing: {:?}. Reconnecting...", e);
```

Comment:
I think that we should add some retry policy here. Without that, reconnecting can fall into the infinite loop and start eating all resources. When the database is down for a longer time, it can kill the client's instance. E.g. in the Kubernetes deployment, this may cause a chain reaction.
I think that a must-have is to give the user the possibility to set the maximum number of retries. It would be good to implement exponential backoff for the increasing timeout between the retries.
I'm not sure how it's in Rust, but in .NET there is a Polly library that supports those scenarios out of the box https://github.com/App-vNext/Polly.

Reply:
This is actually what is happening. `error!` is just for the log. it keeps looping until we managed to connect successfully. As discussed, I'm going to expose settings to limit how many time we should try before giving up.